pkg/schema/definitions/handler.go

```diff
@@ -0,0 +1,197 @@
+	models *proto.Models
+	// schemaToModel is a map of the schema name to the model for that schema. Can be used to load the
+	// top-level definition for a schema, which can then be processed by the schemaFieldVisitor.
+	schemaToModel map[string]string
+}
+
+// byIDHandler is the Handler method for a request to get the schema definition for a specifc schema. Will use the
+// cached models found during the last refresh as part of this process.
+func (s *schemaDefinitionHandler) byIDHandler(request *types.APIRequest) (types.APIObject, error) {
+	// pseudo-access check, designed to make sure that users have access to the schema for the definition that they
+	// are accessing.
+	requestSchema := request.Schemas.LookupSchema(request.Name)
+	if requestSchema == nil {
+		return types.APIObject{}, apierror.NewAPIError(validation.NotFound, "no such schema")
+	}
+
+	if s.needsRefresh() {
+		err := s.refresh()
+		if err != nil {
+			logrus.Errorf("error refreshing schemas %s", err.Error())
+			return types.APIObject{}, apierror.NewAPIError(internalServerErrorCode, "error refreshing schemas")
+		}
+	}
+
+	// lock only in read-mode so that we don't read while refresh writes. Only use a read-lock - using a write lock
+	// would make this endpoint only usable by one caller at a time
+	s.RLock()
+	defer s.RUnlock()
+
+	if s.models == nil {
+		return types.APIObject{}, apierror.NewAPIError(notRefreshedErrorCode, "schema definitions not yet refreshed")
+	}
+	models := *s.models
+	modelName, ok := s.schemaToModel[requestSchema.ID]
+	if !ok {
+		return types.APIObject{}, apierror.NewAPIError(notRefreshedErrorCode, "no model found for schema, try again after refresh")
+	}
+	model := models.LookupModel(modelName)
+	protoKind, ok := model.(*proto.Kind)
+	if !ok {
+		errorMsg := fmt.Sprintf("model for %s was type %T, not a proto.Kind", modelName, model)
+		return types.APIObject{}, apierror.NewAPIError(internalServerErrorCode, errorMsg)
+	}
+	definitions := map[string]definition{}
+	visitor := schemaFieldVisitor{
+		definitions: definitions,
+		models:      models,
+	}
+	protoKind.Accept(&visitor)
+
+	return types.APIObject{
+		ID:   request.Name,
+		Type: "schemaDefinition",
+		Object: schemaDefinition{
+			DefinitionType: modelName,
+			Definitions:    definitions,
+		},
+	}, nil
+}
+
+// needsRefresh readLocks and checks if the cache needs to be refreshed.
+func (s *schemaDefinitionHandler) needsRefresh() bool {
+	s.RLock()
+	defer s.RUnlock()
+	if s.lastRefresh.IsZero() {
+		return true
+	}
+	return s.lastRefresh.Add(s.refreshStale).Before(time.Now())
+}
+
+// refresh writeLocks and updates the cache with new schemaDefinitions. Will result in a call to kubernetes to retrieve
+// the openAPI schemas.
+func (s *schemaDefinitionHandler) refresh() error {
+	s.Lock()
+	defer s.Unlock()
+	openapi, err := s.client.OpenAPISchema()
+	if err != nil {
+		return fmt.Errorf("unable to fetch openapi definition: %w", err)
+	}
+	models, err := proto.NewOpenAPIData(openapi)
```

Comment:
I would not expect this call to be very CPU heavy. If you have a specific test that you would like me to run let me know.

The main purpose of the cache is to prevent the extra network call, which is expensive due to the sheer volume of data being transferred.

Reply:
I would like to take a look at a flame graph once a development image is ready. Nothing blocking this PR